action/action_manager_test.go

```diff
@@ -0,0 +1,179 @@
+	}
+	return []parade.OwnedTaskData{}, nil
+}
+
+func (m mockParade) ReturnTask(_ parade.TaskID, _ parade.PerformanceToken, _ string, _ parade.TaskStatusCodeValue) error {
+	atomic.AddInt32(m.returnCalled, 1)
+	return nil
+}
+
+type mockHandler struct {
+	handleCalled *int32
+}
+
+func newMockHandler() mockHandler {
+	handleCalled := int32(0)
+	return mockHandler{
+		handleCalled: &handleCalled,
+	}
+}
+
+func (m mockHandler) Handle(_ string, _ *string) HandlerResult {
+	atomic.AddInt32(m.handleCalled, 1)
+	return HandlerResult{}
+}
+
+func (m mockHandler) Actions() []string {
+	return []string{"one", "two"}
+}
+
+func (m mockHandler) Actor() parade.ActorID {
+	return "mock"
+}
+
+func durationPointer(d time.Duration) *time.Duration {
+	return &d
+}
+
+func TestManager(t *testing.T) {
+	tests := []struct {
+		name                    string
+		mp                      mockParade
+		mh                      mockHandler
+		sleepTime               time.Duration
+		expectedReturnTaskCalls int32
+		expectedHandleCalled    int32
+		properties              *Properties
+	}{
+		{
+			name:                    "no tasks",
+			mp:                      newMockParade([]ownTaskResult{}),
+			mh:                      newMockHandler(),
+			sleepTime:               100 * time.Millisecond,
+			expectedReturnTaskCalls: int32(0),
+			expectedHandleCalled:    int32(0),
+			properties:              nil,
+		},
+		{
+			name: "50 tasks in one call",
+			mp: newMockParade([]ownTaskResult{
+				{
+					tasks:  50,
+					action: "first action",
+				},
+			}),
+			mh:                      newMockHandler(),
+			sleepTime:               100 * time.Millisecond,
+			expectedReturnTaskCalls: int32(50),
+			expectedHandleCalled:    int32(50),
+			properties:              nil,
+		},
+		{
+			name: "80 tasks in two calls",
+			mp: newMockParade([]ownTaskResult{
+				{
+					tasks:  50,
+					action: "first action",
+				},
+				{
+					tasks:  30,
+					action: "second action",
+				},
+			}),
+			mh:                      newMockHandler(),
+			sleepTime:               100 * time.Millisecond,
+			expectedReturnTaskCalls: int32(80),
+			expectedHandleCalled:    int32(80),
+			properties:              nil,
+		},
+		{
+			name: "exit before second call",
+			mp: newMockParade([]ownTaskResult{
+				{
+					tasks:  50,
+					action: "first action",
+				},
+				{
+					tasks:  0,
+					action: "force sleep",
+				},
+				{
+					tasks:  30,
+					action: "second action",
+				},
+			}),
+			mh:                      newMockHandler(),
+			sleepTime:               50 * time.Millisecond,
+			expectedReturnTaskCalls: int32(50),
+			expectedHandleCalled:    int32(50),
+			properties: &Properties{
+				waitTime: durationPointer(time.Millisecond * 100),
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			a := NewAction(tt.mh, tt.mp, tt.properties)
+			time.Sleep(tt.sleepTime)
+			a.Close()
+			returnCalled := atomic.LoadInt32(tt.mp.returnCalled)
```

Comment:
I don't understand why we are using the subtleties of `atomic`.  There are a channel operation and a wait between the write to the variable and this read.  By e.g. https://golang.org/ref/mem#tmp_7 it happens before.  Does the race detector complain about a regular read here?
In any case, without these synchronizations I am not even sure why an atomic load would happen after a regular store.